### DISCUSSION TITLE
PowerShell is considering enabling VT support *unconditionally* in its `conhost.exe` sessions. Is it safe?

### DISCUSSION THREAD

**Discussion by mklement0** [microsoft/terminal#14802] [General]
# Description of the new feature/enhancement

Note: This isn't a feature request, but since it is also isn't a bug report, I chose this issue type.

PowerShell (Core) currently selectively _disables_ VT support for calls to _external programs_ when running in `conhost.exe` windows - _unless_ the `VirtualTerminalLevel` value in `HKEY_CURRENT_USER\Console` exists and is nonzero (this value doesn't exist by default) - while _enabling_ support for PowerShell-native commands.

This makes for an awkward asymmetry:

* While``"It ain't easy being `e[32mgreen`e[m."`` (a PowerShell-native command) properly renders the VT sequences,
* ``python -c 'print(''It ain\''t easy being \x1b[32mgreen\x1b[m.'')'`` does _not_ (unless explicitly enabled via the registry)

This is non-obvious, given that the exact same string is emitted in both cases.
Currently, unless VT support is explicitly enabled via the registry, the garbled `It ain't easy being ←[32mgreen←[m.` prints to the console.

By contrast, _Windows Terminal_ already has VT support enabled _unconditionally_.

https://github.com/PowerShell/PowerShell/issues/19101 proposes making PowerShell (specifically, transiently, for its own processes and its child processes only) similarly enable VT support _unconditionally_.

@SteveL-MSFT wonders whether it is safe to do so: 

* Does anyone anticipate any problems if this were to be done?

**Comment by DHowett** [comment]
I moved this over to discussions! Sorry, we don't have clear signposts about what does where. 😆 

Once I'm back inside work hours, I'd love to write up the pitfalls we've run into doing this... and what we're looking at doing to improve the situation going forward. I've added a TODO item!

**Reply by floh96** [reply]
@DHowett Have you had time to look at this?

**Comment by mklement0** [comment]
Looking forward to the write-up.

@o-sdn-o has found this comment of yours from a while back: https://github.com/microsoft/terminal/issues/6634#issuecomment-647726408, which - if I understand correctly - implies that Windows Terminal has already decided to break backward compatibility with legacy CLIs that rely on the ability to  "print raw control characters directly to the screen", correct?

**Reply by iSazonov** [reply]
> Windows Terminal has already decided to break backward compatibility

This is manifested on main page https://github.com/microsoft/terminal#the-windows-console-host
> However, because Windows Console's primary goal is to maintain backward compatibility, we have been unable to add many of the features the community (and the team) have been wanting for the last several years including tabs, unicode text, and emoji.

> These limitations led us to create the new Windows Terminal.

**Comment by mklement0** [comment]
Note that it also states, with respect to `conhost.exe`:

"Since taking ownership of the Windows command-line in 2014, the team added several new features to the Console, including background transparency, line-based selection, **support for [ANSI / Virtual Terminal sequences](https://en.wikipedia.org/wiki/ANSI_escape_code), [24-bit color](https://devblogs.microsoft.com/commandline/24-bit-color-in-the-windows-console/)**, a [Pseudoconsole ("ConPTY")](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/), and more."

That, combined with the quotes from my previous comment, suggests that it is ultimately `conhost.exe` providing the VT rendering _both_ in console windows and Windows Terminal.

So the question for PowerShell comes down to:

* Is not supporting legacy CLIs anymore that rely on printing raw control sequences a real-world problem in _PowerShell_, specifically, or does it constitute a beneficial [bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) change?

  * Running such CLIs without a shell or via `cmd.exe` will still be compatible.

* I don't know the answer, and, at least pre-Windows 11,  "just use Windows Terminal" is probably not a viable solution, because you don't always get to control how a console is created.

**Comment by mklement0** [comment]
(On a side note, it seems that when I post comments they are auto-up-voted, ostensibly by me - strange...)

**Reply by zadjii-msft** [reply]
I believe that's just how GitHub Discussions works, kinda like reddit in that sense
